mne/epochs.py

```diff
@@ -2064,6 +2064,280 @@ def _drop_log_stats(drop_log, ignore=('IGNORED',)):
     return perc
 
 
+def make_metadata(events, event_id, tmin, tmax, sfreq,
+                  time_locked_events=None, keep_first=None, keep_last=None):
+    """Generate a default set of `~mne.Epochs` metadata.
+
+    This function generates metadata based on all events falling into the time
+    interval ``[tmin, tmax]``. Only events specified in ``event_id`` will be
+    considered.
+
+    The function will also return a subset of the events array that corresponds
+    to the generated metadata, so both the metadata and the events may directly
+    be used to create `~mne.Epochs`.
+
+    You may specify which events are the time-locked events that will later
+    be used to create `~mne.Epochs` via the ``time_locked_events`` parameter.
+
+    By default, the time of the **first occurrence** of each event is extracted
+    from every time period. To keep the **last occurrence** of a certain
+    event instead, pass ``keep_last``.
+
+    Lastly, you may specify hierarchical event descriptors (HED) in
+    ``keep_first`` and ``keep_last`` to keep the time of the first and last
+    occurrence, respectively, of an entire group of different events.
+    For example, if you have the following hierarchical event names:
+    ``stim/a``, ``stim/b`` and pass ``keep_first='stim'``, the time of
+    **either** of the two events will be extracted, depending on which of the
+    two occurred first in the current time period.
```

Comment:
I know you said I should look at the tutorial first, but I'm not doing that: I should be able to read the docstring and basically grasp how this works.  I tried that and failed, so what follows is me thinking out loud about what it might mean, in order to hopefully point the way to a better explanation / organization of this info.

1. The very first line says it generates Epochs metadata, so I'm surprised to realize it doesn't take an Epochs object.
2. I can't tell: what exactly is "default" about the metadata that it generates?
3. are `tmin`/`tmax` relative to something?  since you're not passing Raw or Epochs, I guess they must be relative to `time_locked_events`?
4. OK, I'm starting to understand I think:  You provide an events array, specify which Event IDs you intend to later use to make epochs (this is `time_locked_events`, right?), and you specify an `event_id` dict and a `tmin`/`tmax` to determine what *other* events to care about, and then this function does some magic and returns a smaller events array (to pass to epochs constructor) and a dataframe with some info about all those other events in it.
5. Early on it says it uses *all* events but later you talk about only using the first (or last) event between `tmin` and `tmax`.  So it doesn't actually use all events?
6. Not clear to me yet whether `keep_first` and `keep_last` are mutually exclusive...  I suppose I could `keep_first='buttonpress'` and `keep_last='screen_flip'` or something?

Reply:
> 4\. OK, I'm starting to understand I think:  You provide an events array, specify which Event IDs you intend to later use to make epochs (this is `time_locked_events`, right?), and you specify an `event_id` dict and a `tmin`/`tmax` to determine what _other_ events to care about, and then this function does some magic and returns a smaller events array (to pass to epochs constructor) and a dataframe with some info about all those other events in it.

Yes, that's precisely what happens 😅

> 6\. Not clear to me yet whether `keep_first` and `keep_last` are mutually exclusive...  I suppose I could `keep_first='buttonpress'` and `keep_last='screen_flip'` or something?

Yes, your example is entirely possible with the current implementation! The function just ensures that users cannot do `keep_first='button', keep_last='button'`.

Reply:
> 1\. The very first line says it generates Epochs metadata, so I'm surprised to realize it doesn't take an Epochs object.

I've tried to address this one now too.

Reply:
> 2\. I can't tell: what exactly is "default" about the metadata that it generates?

I had tried to outline this in the `Returns` section for the `metadata` return value, did you not see this or do you think this info is placed or written badly there?